fe/fe-core/src/main/java/com/starrocks/sql/optimizer/QueryMaterializationContext.java

```diff
@@ -0,0 +1,104 @@
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.rewrite.ScalarOperatorRewriter;
+import com.starrocks.sql.optimizer.rule.transformation.materialization.PredicateSplit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Set;
+
+/**
+ * Store materialized view context during the query lifecycle which is seperated from per materialized view's context.
+ */
+public class QueryMaterializationContext {
+    protected static final Logger LOG = LogManager.getLogger(QueryMaterializationContext.class);
+
+    // Because PredicateSplit's construct is expensive, cache query predicates to its final predicate split per query.
+    // Cache query predicate to its canonized predicate to avoid one predicate's repeat canonized.
+    private final Cache<Object, Object> mvQueryContextCache = Caffeine.newBuilder()
+            .maximumSize(Config.mv_query_context_cache_max_size)
+            .recordStats()
+            .build();
+
+    public QueryMaterializationContext() {
+    }
+
+    public Cache<Object, Object> getMvQueryContextCache() {
+        return mvQueryContextCache;
+    }
+
+    public PredicateSplit getPredicateSplit(Set<ScalarOperator> predicates,
+                                            ReplaceColumnRefRewriter columnRefRewriter) {
+        // Cache predicate split for predicates because it's time costing if there are too many materialized views.
+        Object cached = mvQueryContextCache.getIfPresent(predicates);
+        if (cached != null) {
+            return (PredicateSplit) cached;
+        }
+        ScalarOperator queryPredicate = rewriteOptExprCompoundPredicate(predicates, columnRefRewriter);
+        PredicateSplit predicateSplit = PredicateSplit.splitPredicate(queryPredicate);
+        if (predicateSplit != null) {
+            mvQueryContextCache.put(predicates, predicateSplit);
+        }
+        return predicateSplit;
+    }
+
+    private ScalarOperator rewriteOptExprCompoundPredicate(Set<ScalarOperator> conjuncts,
+                                                           ReplaceColumnRefRewriter columnRefRewriter) {
+        if (conjuncts == null || conjuncts.isEmpty()) {
+            return null;
+        }
+        ScalarOperator compoundPredicate = Utils.compoundAnd(conjuncts);
+        compoundPredicate = columnRefRewriter.rewrite(compoundPredicate.clone());
+        return getCanonizedPredicate(compoundPredicate);
+    }
+
+    public ScalarOperator getCanonizedPredicate(ScalarOperator predicate) {
+        try {
+            return (ScalarOperator) mvQueryContextCache.get(predicate, x -> {
+                ScalarOperator rewritten = new ScalarOperatorRewriter()
+                        .rewrite(predicate.clone(), ScalarOperatorRewriter.MV_SCALAR_REWRITE_RULES);
+                return rewritten;
+            });
+        } catch (NullPointerException e) {
```

Comment:
why catch NPE? when throw NPE ?

Reply:
fix it. Caffeine cache will support npe when key or value is null.